File: packages/engine/src/game.ts
/** SPEC §6 — the game lifecycle. */
import {
  type Board,
  type Direction,
  type Spawn,
  DIRECTION_LETTERS,
  boardToHex,
  emptyBoard,
  isOver,
  maxTile,
  moveInPlace,
  spawn,
} from './board.ts';
import { boardHash, hashHex, historyStep } from './hash.ts';
import { Rng, type RngState } from './rng.ts';

export const SPEC_VERSION = 1;

export interface GameSnapshot {
  seed: number;
  board: string;
  score: number;
  moveCount: number;
  maxTile: number;
  over: boolean;
  historyHash: string;
}

export interface StepResult {
  dir: Direction;
  gained: number;
  spawn: Spawn | null;
}

export class Game {
  readonly seed: number;
  board: Board;
  score = 0;
  moveCount = 0;
  private rng: Rng;
  private hash: number;
  private moveLetters: string[] = [];

  constructor(seed: number) {
    this.seed = seed >>> 0;
    this.rng = Rng.fromSeed(this.seed);
    this.board = emptyBoard();
    spawn(this.board, this.rng);
    spawn(this.board, this.rng);
    this.hash = boardHash(this.board);
  }

  /** Apply a move. Returns null (and changes nothing) if the move is invalid. */
  apply(dir: Direction): StepResult | null {
    const gained = moveInPlace(this.board, dir);
    if (gained < 0) return null;
    this.score += gained;
    this.moveCount += 1;
    const s = spawn(this.board, this.rng);
    this.hash = historyStep(this.hash, this.board, dir);
    this.moveLetters.push(DIRECTION_LETTERS[dir]);
    return { dir, gained, spawn: s };
  }

  get over(): boolean {
    return isOver(this.board);
  }

  get historyHash(): string {
    return hashHex(this.hash);
  }

  get moves(): string {
    return this.moveLetters.join('');
  }

  rngState(): RngState {
    return this.rng.state();
  }

  snapshot(): GameSnapshot {
    return {
      seed: this.seed,
      board: boardToHex(this.board),
      score: this.score,
      moveCount: this.moveCount,
      maxTile: maxTile(this.board),
      over: this.over,
      historyHash: this.historyHash,
    };
  }
}

File: engines/rust/src/game.rs
//! SPEC §6 — the game lifecycle.
use serde::Serialize;

use crate::board::{
    board_to_hex, is_over, max_tile, move_board, spawn, Board, Spawn, DIRECTION_LETTERS,
};
use crate::hash::{board_hash, hash_hex, history_step};
use crate::rng::Rng;

pub const SPEC_VERSION: u64 = 1;

/// Serialisable game state (the replay `final` object).
#[derive(Serialize, Clone, Debug, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct Snapshot {
    pub board: String,
    pub score: u64,
    pub move_count: u64,
    pub max_tile: u64,
    pub over: bool,
    pub history_hash: String,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct Step {
    pub dir: usize,
    pub gained: u64,
    pub spawn: Option<Spawn>,
}

#[derive(Clone, Debug)]
pub struct Game {
    pub seed: u32,
    pub board: Board,
    pub score: u64,
    pub move_count: u64,
    rng: Rng,
    hash: u32,
    moves: String,
}

impl Game {
    pub fn new(seed: u32) -> Game {
        let mut rng = Rng::from_seed(seed);
        let mut board: Board = [0u8; 16];
        spawn(&mut board, &mut rng);
        spawn(&mut board, &mut rng);
        let hash = board_hash(&board);
        Game {
            seed,
            board,
            score: 0,
            move_count: 0,
            rng,
            hash,
            moves: String::new(),
        }
    }

    /// Apply a move. Returns `None` (and changes nothing) if the move is invalid.
    pub fn apply(&mut self, dir: usize) -> Option<Step> {
        if dir > 3 {
            return None;
        }
        let r = move_board(&self.board, dir);
        if !r.changed {
            return None;
        }
        self.board = r.board;
        self.score += r.gained;
        self.move_count += 1;
        let s = spawn(&mut self.board, &mut self.rng);
        self.hash = history_step(self.hash, &self.board, dir);
        self.moves.push(DIRECTION_LETTERS[dir]);
        Some(Step {
            dir,
            gained: r.gained,
            spawn: s,
        })
    }

    pub fn over(&self) -> bool {
        is_over(&self.board)
    }

    pub fn history_hash_value(&self) -> u32 {
        self.hash
    }

    pub fn history_hash(&self) -> String {
        hash_hex(self.hash)
    }

    /// One letter per applied move.
    pub fn moves(&self) -> &str {
        &self.moves
    }

    pub fn rng_state(&self) -> [u32; 4] {
        self.rng.state()
    }

    pub fn snapshot(&self) -> Snapshot {
        Snapshot {
            board: board_to_hex(&self.board),
            score: self.score,
            move_count: self.move_count,
            max_tile: max_tile(&self.board),
            over: self.over(),
            history_hash: self.history_hash(),
        }
    }
}
